Fix 48403 - Assertion failed 'call->gtControlExpr == nullptr'

briansull opened the pull request:
In method `fgMorphCopyBlock` don't call `gtCloneExpr` on the `addrSpill`
We also don't call `fgMorphTree` on the `addrSpill` tree as well

briansull commented:
/azp run runtime-coreclr jitstress

briansull commented:
/azp run runtime-coreclr jitstress

AndyAyersMS commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -8128,9 +8129,21 @@ GenTreeCall* Compiler::gtCloneExprCallHelper(GenTreeCall* tree, unsigned addFlag
+    //
+    if (fgGlobalMorph && tree->IsExpandedEarly() && tree->IsVirtualVtable())
+    {
+        // The 'copy' node will create its own expanded sequence when we morph this cloned node
```
Presumably `gtControlExpr` is `nullptr` here? Could we check for that instead?

briansull replied:
This method gets called from  fgMorphCopyBlock (and several other places)

And when called from fgMorphCopyBlock the gtControlExpr has already been expanded as we start from the bottom of the tree and morph upwards.  It also could be nullptr if we are planning on expanding this callsite late.

AndyAyersMS replied:
I guess I don't understand why this needs to be dependent on `fgGlobalMorph` and can't simply be
```
    if (tree->gtControlExpr == nullptr)
```

briansull replied:
Because for some calls we want to expand Virtual Calls late in Lower and for the others we will expand them earlier during Morph. We could revisit this decision, but originally it was pretty simple to implement this way, and I wanted to be able to turn off this change and revert back to the old way to allow us to investigate any bugs that might come up due to this change.
 
By checking `(tree->gtControlExpr == nullptr)` this would match all the cases where we want to expand the call late in Lower, so instaed we always need to check `tree->IsExpandedEarly() && tree->IsVirtualVtable()`. 

Here I also have to check `fgGlobalMorph`, since morph is called multiple times when we optimize (in CSE, etc..), only during the first "global morph" phase do we want to expand the gtControlExpr.  

On these last couple of issues that I am fixing, the problem is that for some unusual code paths, we weren't calling  gtMorphTree at all while in the "global morph" phase.
Thus the `gtControlExpr` was not expanded and the call flags were set to tell Lower that it should not expand `gtControlExpr` as it should have been expanded early.

briansull replied:
Oh Joy,
Now when I add the call to fgMorpgTree of the cloned tree for `STRESS_CLONE_EXPR `it now fails for asmall number of methods with a  `'!"Missing flags on tree"'` assertion cause by `optAssertionProp_Ind` setting the GTF_ORDER_SIDEEFF flag and not propagating it up the tree.

This only happens for a couple of methods when JIitStress is on.

AndyAyersMS replied:
Sorry harp on this, but are you saying if we get here and we're in global morph and are cloning an expand early call and `gtControlExpr` is non-null, we don't clone the existing tree because we know we'll create a proper new tree shortly..?

Won't that just create a tree that's more or less identical to the one we'd get it we did clone the existing tree?

briansull replied:
I will investigate doing it that way.

briansull replied:
That is what we had previously and it causes the assert in the Issue #48403
In he field by field copy path we call fgMorphTree on a cloned tree.

I will try a different fix.

AndyAyersMS reviewed (commented):
The morph changes seem unrelated to fixing the assert. Should they be PR'd separately?

briansull commented:
/azp run runtime-coreclr jitstress

briansull commented:
/azp run runtime-coreclr jitstress

briansull commented:
/azp run runtime-coreclr jitstress

briansull commented:
This is ready for checkin now.
@AndyAyersMS @sandreenko PTAL

briansull commented:
/azp run runtime-coreclr jitstress

briansull committed:
New fix for issue 48403
 - Don't use gtCloneExpr on trees that have persistent side effects
 - Don't call gtCloneTree or fgMorpgTree when we have an addrSpill case

briansull commented:
/azp run runtime-coreclr jitstress

briansull commented:
I have reworked this fix.
It is now a minimal fix that fixes the issue.

- In method `fgMorphCopyBlock` don't call `gtCloneExpr` on the `addrSpill`
- We also don't call `fgMorphTree` on the `addrSpill` tree as well

This is ready for checkin now.
@AndyAyersMS @sandreenko PTAL

AndyAyersMS reviewed (approved):
LGTM.

Thanks for reconsidering how to fix this.

briansull commented:
The only failures that I see in the JitStress runs are due to
`eventpipe\eventsvalidation\ThreadPool\ThreadPool.cmd`

Which is a currently known issue:  #48727
